scarb/src/compiler/incremental/source.rs

```
@@ -0,0 +1,109 @@
+            if !p.is_dir() {
+                p.parent().expect("source path must have a parent")
+            } else {
+                p
+            }
+        })
+        .dedup()
+        .collect_vec();
+    let Some((first, rest)) = source_paths.split_first() else {
+        return Vec::new();
+    };
+    let mut builder = WalkBuilder::new(first);
+    for path in rest {
+        builder.add(path);
+    }
+    let mut types_builder = TypesBuilder::new();
+    types_builder
+        .add(CAIRO_FILE_EXTENSION, &format!("*.{CAIRO_FILE_EXTENSION}"))
+        .unwrap();
+    types_builder.select(CAIRO_FILE_EXTENSION);
+    let walker = builder
+        .follow_links(true)
+        .standard_filters(false)
+        .parents(false)
+        .same_file_system(false)
+        .filter_entry(filter)
+        .skip_stdout(true)
+        .types(types_builder.build().unwrap())
+        .build_parallel();
+
+    let (tx, rx) = std::sync::mpsc::channel();
+
+    walker.run(|| {
+        let tx = tx.clone();
+        Box::new(move |result| {
+            let dir_entry = if let Ok(dir_entry) = result {
+                dir_entry
+            } else {
+                warn!("failed to read the file");
+                return Continue;
+            };
+
+            let file_type = if let Some(file_type) = dir_entry.file_type() {
+                file_type
+            } else {
+                warn!("failed to read filetype");
+                return Continue;
+            };
+
+            let path = dir_entry.path();
+
+            if !file_type.is_file() {
+                return Continue;
+            }
+
+            let Ok(path) = path.try_to_utf8() else {
+                warn!("failed to convert path to UTF-8: {}", path.display());
+                return Continue;
+            };
+
+            let Ok(content) = fsx::read_to_string(&path) else {
+                warn!("failed to read file: {}", path.to_string());
+                return Continue;
+            };
```

Did you consider what happens if, for example, there's a permission issue with the file, the file is corrupted, or the file got deleted before the walker runs?

While some of these are more and less realistic, I worry that silently ignoring such failures with `Continue` may produce stale cache files 🤔 

I wonder if it make sense to instead produce `Error` in such circumstance, display terminal warning, and treat such occurrence as cache miss.
 
WDYT?

---

Good question

Generally, I don't think it is completely safe to turn problems here into hard errors either, as we traverse any files under `/src` dir -  whether they are part of the package source code or not. At this point in compilation, we have no way of knowing, which files will be part of the module tree - determined by `mod` keywords basically. 

> Did you consider what happens if, for example, there's a permission issue with the file, the file is corrupted, or the file got deleted before the walker runs?

In a sense, this does not need to be a problem for us: 

If a file has been deleted before run -> it has been previously in the hash and now will be skipped -> hash will change -> cache will be overwritten. 

If a file is corrupt / has permissions that deny us from accessing it -> we won't be able to load it during compilation neither. (and if it has been accessible before, the hash will change, like in the previous example). 

It's possible that there is some edge case where it is a problem, but not sure if we have a better way of handling this 🤔

---

Yeah, I can do that. Changed!